docs/how-to-guides/upgrade-openebs.md

```diff
@@ -0,0 +1,278 @@
+
+* Lokomotive `v0.6.0`
+    ```bash
+    lokoctl version
+    v0.6.0
+    ```
+
+* A Lokomotive cluster accessible via `kubectl` deployed on a [Equinix]
+    metal(https://metal.equinix.com/).
+
+* OpenEBS `v1.12.0` installed. You can check if OpenEBS is indeed in expected
+    version:
+    ```bash
+    $ kubectl get pods -n openebs -n openebs.io/version=1.12.0
+    ```
+
+* Highly recommended to schedule a downtime for the applications consuming the
+    OpenEBS PV and make sure to take a backup of the data before starting the
+    below upgrade procedure.
+
+
+## Steps
+
+###Step 1: Upgrade OpenEBS control plane components
+
+Lokomotive provides an easy way of upgrading the OpenEBS control plane.
+
+Execute the following command to upgrade OpenEBS control plane:
+
+```bash
+lokoctl component apply openebs-operator
+```
+
+Doing so, terminates the OpenEBS resources associated with `v1.12.0` and new
+resources are created with `v2.2.0`.
+
+Verify all the pods are in `Running` state before proceeding:
+
+```bash
+$ kubectl get pods -n openebs -n openebs.io/version=2.2.0
+```
+
+### Step 2: Upgrade OpenEBS data plane components
+
+OpenEBS control plane and data plane components work independently. Even if the
+control plane components are upgraded to `v2.2.0`, the data plane components
+continue to work with the older version `v1.12.0`.
+
+#### Upgrade cStor pools
+
+Get existing `StoragePoolClaims` by executing:
+
+```bash
+kubectl get spc
+NAME                           AGE
+cstor-pool-openebs-replica-3   2d21h26m
+```
+
+Create a Job to upgrade the existing cStor pools:
+
+```bash
+# upgrade-cstor-pools-1120-220.yaml
+
+#This is an example YAML for upgrading cstor SPC.
+#Some of the values below needs to be changed to
+#match your openebs installation. The fields are
+#indicated with VERIFY
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  #VERIFY that you have provided a unique name for this upgrade job.
+  #The name can be any valid K8s string for name. This example uses
+  #the following convention: cstor-spc-<flattened-from-to-versions>
+  name: cstor-spc-1120220
+
+  #VERIFY the value of namespace is same as the namespace where openebs components
+  # are installed. You can verify using the command:
+  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
+  # The above command should return status of the openebs-apiserver.
+  namespace: openebs
+spec:
+  backoffLimit: 4
+  template:
+    spec:
+      serviceAccountName: openebs-operator
+      containers:
+      - name:  upgrade
+        args:
+        - "cstor-spc"
+
+        # --from-version is the current version of the pool
+        - "--from-version=1.12.0"
+
+        # --to-version is the version desired upgrade version
+        - "--to-version=2.2.0"
+
+        # Bulk upgrade is supported
+        # To make use of it, please provide the list of SPCs
+        # as mentioned below
+        - "cstor-pool-openebs-replica-3"
```

Comment:
Same for storage pools, we should also be able to automatically obtain this list I think.

Reply:
I don't want to complicate the upgrade steps.

Reply:
IMO it would simplify it.

Reply:
@invidian automatically obtain this list and then how to render ?

Reply:
we already use `cat << EOF` here. we can just replace this with `$()`.

Reply:
ok but then how to get it listed in the below format without reducing readibility.
```
- "cstor-pool-openebs-replica-3"
- "xyz"
- "abc"
```

Reply:
@invidian this works :see_no_evil: but I am not sure if I want that.
```
OPENEBS_OLD_VERSION=1.12.0
OPENEBS_NEW_VERSION=2.2.0
cat > upgrade-cstor-pools-1120-220.yaml <<EOF
# upgrade-cstor-pools-1120-220.yaml

#This is an example YAML for upgrading cstor SPC.
#Some of the values below needs to be changed to
#match your openebs installation. The fields are
#indicated with VERIFY
---
apiVersion: batch/v1
kind: Job
metadata:
  #VERIFY that you have provided a unique name for this upgrade job.
  #The name can be any valid K8s string for name. This example uses
  #the following convention: cstor-spc-<flattened-from-to-versions>
  generateName: cstor-spc-1120220-
  namespace: openebs

spec:
  backoffLimit: 4
  template:
    spec:
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:
        - "cstor-spc"

        # --from-version is the current version of the pool
        - "--from-version=${OPENEBS_OLD_VERSION}"

        # --to-version is the version desired upgrade version
        - "--to-version=${OPENEBS_NEW_VERSION}"

        # Bulk upgrade is supported
        # To make use of it, please provide the list of SPCs
        # as mentioned below
        - "cstor-pool-openebs-replica-3"
$(kubectl get cstorvolumes -n openebs | awk '{if(NR>1) printf("\"%s\"\n", $1);}'| sed 's/./        - &/')


        #Following are optional parameters
        #Log Level
        - "--v=4"
        #DO NOT CHANGE BELOW PARAMETERS
        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        tty: true

        # the image version should be same as the --to-version mentioned above
        # in the args of the job
        image: openebs/m-upgrade:${OPENEBS_NEW_VERSION}
        imagePullPolicy: Always
      restartPolicy: OnFailure
EOF
```

Reply:
Looks okay, but maybe use `cut` or `awk '{print $1}'` with `tr \\n ' '` or something to make it easier to read? I can help if you give me raw output from `kubectl get cstorvolumes -n openebs` :)

Or even better, use `kubectl get cstorvolumes -n openebs --no-headers -o custom-columns=":metadata.name" | sed 's/^/        - /'`